### ISSUE TITLE
Bulk import ignores onDuplicate in 3.8.0

### ISSUE THREAD

**Issue by mirlos-pt** [arangodb/arangodb#14720]
## My Environment

* __ArangoDB Version__:        3.8.0
* __Storage Engine__:             RocksDB
* __Deployment Mode__:        Single Server
* __Deployment Strategy__:   Manual Start in Docker | Kubernetes
* __Configuration__:               N/A
* __Infrastructure__:               own
* __Operating System__:        Ubuntu 20.04
* __Total RAM in your machine__:        1.5Tb
* __Disks in use__:        HDD
* __Used Package__:            Docker - official Docker library

## Component, Query & Data

__Affected feature__:
Bulk import HTTP API with onDuplicate=update/replace

__Dataset__:
```
$ cat /tmp/testnode.txt 
["_key", "name"]
["one", "One"]
["two", "Two"]
["three", "Three"]
```

__Size of your Dataset on disk__:
66b

## Steps to reproduce

```
# docker run -d -p 30800:8529 --name ar30800 -e ARANGO_ROOT_PASSWORD=issue arangodb:3.8
0f80b32924b5a2fef3f8bd51118f603dd487cf3acd0ab97a5c756da040dc6954

# curl -u root:issue -X POST 'http://localhost:30800/_api/database' -d '{"name": "duplicated"}'
{"error":false,"code":201,"result":true}

# curl -u root:issue -X POST 'http://localhost:30800/_db/duplicated/_api/collection' -d '{"name": "nodes"}'
{"error":false,"code":200,"writeConcern":1,"waitForSync":false,"usesRevisionsAsDocumentIds":true,"type":2,"syncByRevision":true,"id":"198","cacheEnabled":false,"isSmartChild":false,"objectId":"197","globallyUniqueId":"hBBF4F9AE755/198","schema":null,"keyOptions":{"allowUserKeys":true,"type":"traditional","lastValue":0},"isSystem":false,"isDisjoint":false,"name":"nodes","statusString":"loaded","status":3}

# curl -u root:issue -X POST 'http://localhost:30800/_db/duplicated/_api/import?collection=nodes&onDuplicate=update&details=true' --data-binary @/tmp/testnode.txt
{"error":false,"created":3,"errors":0,"empty":0,"updated":0,"ignored":0,"details":[]}

# curl -u root:issue -X POST 'http://localhost:30800/_db/duplicated/_api/import?collection=nodes&onDuplicate=update&details=true' --data-binary @/tmp/testnode.txt
{"error":false,"created":0,"errors":3,"empty":0,"updated":0,"ignored":0,"details":["at position 0: creating document failed with error 'unique constraint violated', offending document: {\"_key\":\"one\",\"name\":\"One\"}","at position 1: creating document failed with error 'unique constraint violated', offending document: {\"_key\":\"two\",\"name\":\"Two\"}","at position 2: creating document failed with error 'unique constraint violated', offending document: {\"_key\":\"three\",\"name\":\"Three\"}"]}

# curl -u root:issue -X POST 'http://localhost:30800/_db/duplicated/_api/import?collection=nodes&onDuplicate=replace&details=true' --data-binary @/tmp/testnode.txt
{"error":false,"created":0,"errors":3,"empty":0,"updated":0,"ignored":0,"details":["at position 0: creating document failed with error 'unique constraint violated', offending document: {\"_key\":\"one\",\"name\":\"One\"}","at position 1: creating document failed with error 'unique constraint violated', offending document: {\"_key\":\"two\",\"name\":\"Two\"}","at position 2: creating document failed with error 'unique constraint violated', offending document: {\"_key\":\"three\",\"name\":\"Three\"}"]}
```

__Problem__:

Document updates are rejected despite onDuplicate=update or onDuplicate=replace being set in the request, as per https://www.arangodb.com/docs/stable/http/bulk-imports.html. This is a regression from 3.7.

__Expected result__:

```
# docker run -d -p 30714:8529 --name ar30714 -e ARANGO_ROOT_PASSWORD=issue arangodb:3.7
8f221e013a37fdde1c4ad313c6785d60508da2ac520ae6d20be354c37ed7fca1

# curl -u root:issue -X POST 'http://localhost:30714/_api/database' -d '{"name": "duplicated"}'
{"error":false,"code":201,"result":true}

# curl -u root:issue -X POST 'http://localhost:30714/_db/duplicated/_api/collection' -d '{"name": "nodes"}'
{"error":false,"code":200,"writeConcern":1,"waitForSync":false,"type":2,"tempObjectId":"0","id":"1017","cacheEnabled":false,"isSmartChild":false,"objectId":"1016","globallyUniqueId":"hC8E6C39918F5/1017","schema":null,"keyOptions":{"allowUserKeys":true,"type":"traditional","lastValue":0},"isSystem":false,"isDisjoint":false,"name":"nodes","statusString":"loaded","status":3}

# curl -u root:issue -X POST 'http://localhost:30714/_db/duplicated/_api/import?collection=nodes&onDuplicate=update&details=true' --data-binary @/tmp/testnode.txt
{"error":false,"created":3,"errors":0,"empty":0,"updated":0,"ignored":0,"details":[]}

# curl -u root:issue -X POST 'http://localhost:30714/_db/duplicated/_api/import?collection=nodes&onDuplicate=update&details=true' --data-binary @/tmp/testnode.txt
{"error":false,"created":0,"errors":0,"empty":0,"updated":3,"ignored":0,"details":[]}

# curl -u root:issue -X POST 'http://localhost:30714/_db/duplicated/_api/import?collection=nodes&onDuplicate=replace&details=true' --data-binary @/tmp/testnode.txt
{"error":false,"created":0,"errors":0,"empty":0,"updated":3,"ignored":0,"details":[]}
```

**Comment by dothebart** [comment]
shouldn't the toplevel of a document be `{}` ? 
If you specify [], it will most probably become:
```
{
   "_key": "autogenerated key",
  "...": your imported thing
}
```

**Comment by Simran-B** [comment]
CSV-style arrays should actually work if the `type` parameter is omitted, see https://github.com/arangodb/arangojs/issues/496#issuecomment-358082928

**Comment by maxkernbach** [comment]
Hi @mirlos-pt,

Thank you for reporting this issue. According to our documentation your dataset format is valid using the `_api/import` API.

> The body must consist of JSON-encoded arrays of attribute values, with one line per document. The first row of the request must be a JSON-encoded array of attribute names. These attribute names are used for the data in the subsequent lines.

As you stated, the same use of the bulk import API with `onDuplicate=update/replace` succeeds in latest 3.7 (3.7.14) and fails in latest 3.8 (3.8.1). In particular, the update/replace fails even though the `type` parameter is omitted.

When changing the input format to JSON lines, the option `onDuplicate=update/replace` works as expected in 3.8.

input file:
```
cat /tmp/test2.txt 
{ "_key": "one", "name": "One"}
{ "_key": "two", "name": "Two"}
{ "_key": "three", "name": "Three"}
```

request with `onDuplicate=replace&type=documents`:
```
curl -u root:issue -X POST 'http://localhost:8529/_db/duplicated/_api/import?collection=nodes&onDuplicate=replace&details=true&type=documents' --data-binary @/tmp/test2.txt | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   185  100    85  100   100  14166  16666 --:--:-- --:--:-- --:--:-- 30833
{
  "error": false,
  "created": 0,
  "errors": 0,
  "empty": 0,
  "updated": 3,
  "ignored": 0,
  "details": []

```

**Comment by jsteemann** [comment]
Fixed via the 2 above PRs. Bugfix is supposed to land in 3.8.2.

**Comment by maxkernbach** [comment]
Hi @mirlos-pt,

I am closing this issue since a fix is available since version 3.8.2 (3.8.6 is currently the latest 3.8.x version).
In case you can still reproduce an issue with `onDuplicate=update/replace`, please comment. 

[CHANGELOG](https://github.com/arangodb/arangodb/blob/5a1026a00b77d26568a83b636f75444504df4e3c/CHANGELOG#L421) entry:

https://github.com/arangodb/arangodb/blob/5a1026a00b77d26568a83b636f75444504df4e3c/CHANGELOG#L421
